sisl/io/sile.py

```diff
@@ -546,6 +547,50 @@ def pre_open(self, *args, **kwargs):
     return _wrapper
 
 
+def sile_read_multiple(all=False, start=0, step=None):
```

Comment:
instead of `all` perhaps `start, stop, step`. This goes in line with slice arguments.

Come to think of it, would it make sense to have a single argument named `range` or `rng` this could then also accept slices? Hmm... Might just be annoying when you only need a single element.

Which would you prefer:
```python
# end is not inclusive
read_geometry(range=5) # same as start=0, step=1, stop=5
read_geometry(range=(2, 5)) # same as start=2, step=1, stop=5
read_geometry(range=(2, 5, 2)) # same as start=2, step=2, stop=5
read_geometry(range=slice(2, None)) # same as start=2, stop=-1

# or the more explicit
read_geometry(stop=5)
read_geometry(start=2, stop=6, step=2)
```

Reply:
I see your point, but I thought `all` is more in line with other sile functions, like [`outSileSiesta.read_geometry`](https://zerothi.github.io/sisl/api/io/generated/sisl.io.siesta.outSileSiesta.html#sisl.io.siesta.outSileSiesta.read_geometry).

Reply:
Thanks! Yes, but come to think of it, I think the other way would be more pythonic, then we could change the other ones to follow suit on this convention. What do you think?

Reply:
One could also just stick with `all` (simple and intuitive keyword) and let the user process the returned list if they want only a subset. The whole file will be processed anyway, so I am not sure there is much of an advantage of leaving out data.

Reply:
In some cases reading very large files could be very slow if one is only interested in reading every 5th value. In this case having a `skip` argument to the decorator could faster skip the geometry reading. For instance converting coordinates and atoms into the `Atoms` class + `Geometry` class turns out to be quite heavy when reading, say 1000, geometries.

What about we do this:
```python
def sile_read_multiple(start=0, end=-1, step=None, all=False):
    if all:
        start, end, step = 0, -1, 1
```
this would allow fast argument, and enable the other one as well?

Reply:
Oh, and when not reading everything, it would be wise to discard them on the fly due to the memory overhead, which again isn't big for small MD collections, but for big it could be.

Reply:
OK, good points. Is there a fast way to step forward a number of lines in a file without invoking `readline()` in a `for` loop?

Reply:
No, not really.

The simplest would be to do `for _ in range(N): self.fh.readline()`, that bypasses all parsers etc., i.e. you don't want to call `self.readline()` as that also does some comment parsing etc.

Reply:
One problem is that the number of atoms in each geometry in a series may not need to be the same. To find out one needs the first line.

Would it be a good idea to implement a keyword in `xyzSile.read_geometry(skip=True)` which steps over the geometry data and return, say, only the number of atoms and the comment line?

Reply:
So you mean the skip function needs to adapt, but that skip function will still read `na` and then parse the correct number of lines. Perhaps I am missing something, but I don't see a problem since `read_geometry` is self-contained and will only ever read the number of lines matching the number of atoms + 2. The `skip` should be a callable that simple reads the file, for an xyz sile it could be:
```python
def skip(self):
    na = int(self.readline())
    self.readline()
    for _ in range(na): self.readline()
```

Reply:
> Would it be a good idea to implement a keyword in `xyzSile.read_geometry(skip=True)` which steps over the geometry data and return, say, only the number of atoms and the comment line?

If anything this should be a separate method, but I don't like it ;) I don't think it is necessary?

Reply:
Do you have an example?

Reply:
I don't see how `sile_read_multiple` can know how many lines to skip without actually calling the implemented read function. One only knows this after entering the data block.

Reply:
Ok, perhaps I was not clear. The function should be something like this:
```python
def sile_read_multiple(all=False, start=0, step=None, skip_call=None, pre_call=None, post_call=None):
    """ Method decorator for doing multiple reads
    Parameters
    ----------
    all : bool, optional
       read all entries (takes precedence over start/step)
    start : int, optional
       position of first read
    step : int, optional
       number of steps to take (if any at all)
   *_call : callable, optional
        callables called before/after read and for fast skipping an element

    Returns
    -------
    single entry or list from multiple reads
    """
    def decorator(func):
         if skip_call is None:
              skip_call = func # perhaps this should be partial'ized with arguments... hmm
        @wraps(func)
        def multiple(*args, all=all, start=start, step=step, end=end, **kwargs):
            if all is True:
                start, end, step = 0, 1e100, 1
            rng = range(start, end, step)
            R, r = [], True
            def read_next(ir):
                   if ir in rng:
                       pre_call()
                       res = func(*args, **kwargs)
                       post_call()
                       return res
                   pre_call() ; res = skip_call(); post_call()
                   return res
                   
            if hasattr(self, "fh"):
                ir = 0
                r = read_next(ir)
                while r is not None:
                    R.append(r)
                    ir += 1
                    r = read_next(ir)
                
                return R[rng]

       ...

        return multiple
    return decorator

Reply:
I adopted `stop` instead of `end` as this seems to be the custom in `slice`. Otherwise tried to follow your layout.

Reply:
agreed, thanks for fixing my mistake!

Reply:
How about leaving the function arguments untouched but allowing slicing to create the loop?

```python
# The original function call
geom = xyzSile(f).read_geometry(atoms=...)

# How it works now if you want to read multiple geometries
geoms = xyzSile(f).read_geometry(atoms=..., start=0, stop=30)

# How I propose to separate the loop from the function call.
# Prepare the loop before calling the function
geoms = xyzSile(f).read_geometry[:30](atoms=...)

# "All" would simply be
geoms = xyzSile(f).read_geometry[:](atoms=...)
```

This is similar to dispatchers, where you would do something like `xyzSile(f).apply(0, 30).read_geometry(atoms=...)`.

Reply:
I know preparing the loop with slicing is a bit restrictive (it doesn't allow a `pre_call` for example), but you could go for the `.apply(0, 30, pre_call=...)` version of it if you want to further customize it (?)

Reply:
that seems like a good addition, so I imagine `read_geometry[:30](end=15)` would use `end=15` rather than `30` so it _can_ be overwritten in argument list, no? Once https://github.com/tfrederiksen/sisl/pull/2 is in (or a variant) then it would be great with a PR! :)

Reply:
no the idea is that the `pre_call` is only known at decoration time, a user cannot do a `pre_call` overwrite. So that would be fine.

Reply:
I think we shouldn't complicate things unnecessarily, the `[...]` notation would basically just set the defaults in the call.